### PR TITLE
Support Multiple Fuzzers

### PR DESCRIPTION
**Pull request by breeswish**
## What have you changed? (mandatory)

This PR preliminarily adds support for using more fuzzers to run fuzz tests, including AFL and Honggfuzz, which will be run on our schrodinger platform.

Note:

1. Honggfuzz cannot fuzz targets with the whole TiKV library included (because it is too large). So Honggfuzz is only usable after we separate more components into workspaces. This will be resolved by future PRs.
2. AFL needs initial seeds as inputs, which is not included in this PR. It will be resolved by future PRs as well.

Please read [`fuzz/README.md`](https://sourcegraph.com/github.com/tikv/tikv@3b8737f9cc0d9664c6ce1d0d3bc6d0dd66690162/-/blob/fuzz/README.md?diff=3b8737f9cc0d9664c6ce1d0d3bc6d0dd66690162&utm_source=chrome-extension) for new operation guide to run fuzz tests.

Resolves https://github.com/tikv/tikv/issues/3778

## What are the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

### CODE REVIEW

**Comment by breeswish** [comment]
@overvenus @huachaohuang PTAL

**Comment by breeswish** [comment]
@ice1000 Yes! And only need to write once. Then you can run it using multiple different fuzzers.

File: fuzz/targets/mod.rs [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    let mut cursor = Cursor::new(data);
+    let decimal1 = Decimal::from_f64(cursor.read_as_f64()?)?;
+    let decimal2 = Decimal::from_f64(cursor.read_as_f64()?)?;
+    let _ = fuzz(&decimal1, &decimal2, &mut cursor);
+    let _ = fuzz(&decimal2, &decimal1, &mut cursor);
+    Ok(())
+}
+
+register_fuzz!(
+    fuzz_codec_bytes_encode_bytes,
```
Comment (overvenus) [inline fuzz/targets/mod.rs]: Could you seperate different series of functions to different binarys? By this way, fuzzer can find better inputs.

**Comment by breeswish** [comment]
According to @overvenus 's suggestion, we'd better find a way to separate these fuzzer targets into different binaries, instead of merge them together in a single function. Mark this PR as WIP because there will be more work to do.

**Comment by ice1000** [comment]
Related: #3355

**Comment by breeswish** [comment]
@overvenus @ice1000 Updated to support separating fuzz targets into different binaries to be more friendly to fuzzers. PTAL, thanks!

**Comment by breeswish** [comment]
Cargo.lock conflicted again 🙃 let me fix it

File: fuzz/cli.rs [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+         -C debug-assertions=on \
+         -C debuginfo=0 \
+         -C opt-level=3 \
+         -Z sanitizer=address",
+    );
+
+    let mut asan_options = env::var("ASAN_OPTIONS").unwrap_or_default();
+    asan_options.push_str(" detect_odr_violation=0");
+
+    let fuzzer_bin = Command::new("cargo")
```
Comment (overvenus) [inline fuzz/cli.rs]: Why not `cargo-fuzz`? So we dont need to maintain magic rust flags.

Comment (breeswish) [reply fuzz/cli.rs]: cargo-fuzz requires magic Cargo.toml and a magic directory. No way to be good organized.

**Comment by ice1000** [comment]
Resolve conflicts?

**Pull request merged**